server/jetstream_errors.go

```diff
@@ -0,0 +1,220 @@
+package server
+
+import (
+	"fmt"
+)
+
+type APIErrorIdentifier int
+
+const (
+	JSUnknownErr APIErrorIdentifier = iota
+	JSAccountResourcesExceededErr
+	JSBadRequestErr
+	JSClusterIncompleteErr
+	JSClusterNoPeersErr
+	JSClusterNotActiveErr
+	JSClusterNotAssignedErr
+	JSClusterNotAvailErr
+	JSClusterNotLeaderErr
+	JSClusterRequiredErr
+	JSClusterTagsErr
+	JSConsumerCreateErrF
+	JSConsumerNameExistErr
+	JSConsumerNotFoundErr
+	JSDeliverSubjectInvalidErr
+	JSDurableNameNotInSubjectErr
+	JSDurableNameNotMatchSubjectErr
+	JSDurableNameNotSetErr
+	JSEphemeralWithDurableInSubjectErr
+	JSEphemeralWithDurableNameErr
+	JSExternalApiOverlapErrF
+	JSExternalDelPrefixOverlapsErrF
+	JSInsufficientResourcesErr
+	JSInvalidExternalDeliverySubjErrF
+	JSInvalidJSONErr
+	JSMaximumConsumersLimitErr
+	JSMaximumStreamsLimitErr
+	JSMemoryResourcesExceededErr
+	JSMirrorConsumerSetupFailedErrF
+	JSMirrorMaxMessageSizeTooBigErr
+	JSMirrorWithSourcesErr
+	JSMirrorWithStartSeqAndTimeErr
+	JSMirrorWithSubjectFiltersErr
+	JSMirrorWithSubjectsErr
+	JSNoAccountErr
+	JSNoClusterSupportErr
+	JSNoMessageFoundErr
+	JSNotEmptyRequestErr
+	JSNotEnabledErr
+	JSPeerNotMemberErr
+	JSRaftGeneralErrF
+	JSRestoreSubscribeFailedErr
+	JSSequenceNotFoundErrF
+	JSServerNotMemberErr
+	JSSourceConsumerSetupFailedErrF
+	JSSourceMaxMessageSizeTooBigErr
+	JSStorageResourcesExceededErr
+	JSStreamAssignmentErrF
+	JSStreamCreateErrF
+	JSStreamDeleteErrF
+	JSStreamGeneralErrorF
+	JSStreamInvalidConfigF
+	JSStreamLimitsErrF
+	JSStreamMessageExceedsMaximumErr
+	JSStreamMirrorNotUpdatableErr
+	JSStreamMismatchErr
+	JSStreamMsgDeleteFailedF
+	JSStreamNameExistErr
+	JSStreamNotFoundErr
+	JSStreamNotMatchErr
+	JSStreamReplicasNotUpdatableErr
+	JSStreamRestoreErrF
+	JSStreamSequenceNotMatchErr
+	JSStreamSnapshotErrF
+	JSStreamSubjectOverlapErr
+	JSStreamTemplateCreateErrF
+	JSStreamTemplateDeleteErrF
+	JSStreamTemplateNotFoundErr
+	JSStreamUpdateErrF
+	JSStreamWrongLastMsgIDErrF
+	JSStreamWrongLastSequenceErrF
+	JSTempStorageFailedErr
+	JSTemplateNameNotMatchSubjectErr
+)
+
+var (
+	ApiErrors = map[APIErrorIdentifier]*ApiError{
+		JSUnknownErr:                       {Code: 503, ErrCode: "JS001", Description: "Unknown error"},
```

Comment:
Not decided yet if `JS001` is the right format, feedback welcome, I do think strings not numbers are better and the catagorisation that the `JS` adds is nice.

Reply:
Numbers are directly comparable in all languages, that could be an advantage. Interested in what other's think.

Reply:
Numbers would be easier indeed. But if the category is useful (JS), it could be made an ErrCategory: "JS" for people who care to know. But the ErrCode alone (a number) would be enough to check for the error code. However, for applications to be able to ignore the category means that error codes should be ever increasing, that is, do not repeat err codes because they belong to different categories.

PS: I get that Code already exists in ApiError, and RI proposed Kind first. But ErrCode is a bit of a stutter, as in ApiError.ErrCode.

Reply:
See comments below about how to check for errors etc, but I agree probably `JS001` is not needed.  How would you feel about codes like `1xxxx` being JS and `2xxxx` being MQTT etc?